test/cpp/end2end/crl_end2end_test.cc

```diff
@@ -0,0 +1,166 @@
+    certificate_file = absl::StrCat(kCredentialsDir, "/revoked.pem");
+    key_file = absl::StrCat(kCredentialsDir, "/revoked.key");
+  } else {
+    certificate_file = absl::StrCat(kCredentialsDir, "/valid.pem");
+    key_file = absl::StrCat(kCredentialsDir, "/valid.key");
+  }
+  const std::string ca_bundle_file = absl::StrCat(kCredentialsDir, "/ca.pem");
+
+  auto certificate_provider = std::make_shared<FileWatcherCertificateProvider>(
+      key_file, certificate_file, ca_bundle_file,
+      /*refresh_interval_sec=*/10);
+  TlsChannelCredentialsOptions options;
+  options.set_certificate_provider(certificate_provider);
+  options.watch_root_certs();
+  options.watch_identity_key_cert_pairs();
+  options.set_server_verification_option(GRPC_TLS_SKIP_HOSTNAME_VERIFICATION);
+  std::shared_ptr<TestTlsServerAuthorizationCheck>
+      test_server_authorization_check(new TestTlsServerAuthorizationCheck());
+  std::shared_ptr<TlsServerAuthorizationCheckConfig>
+      server_authorization_check_config(new TlsServerAuthorizationCheckConfig(
+          test_server_authorization_check));
+  options.set_server_authorization_check_config(
+      server_authorization_check_config);
+  auto channel_creds = grpc::experimental::TlsCredentials(options);
+  grpc::ChannelArguments args;
+  if (revoked_server_certs) {
+    args.SetString(GRPC_SSL_TARGET_NAME_OVERRIDE_ARG, "revoked");
+  } else {
+    args.SetString(GRPC_SSL_TARGET_NAME_OVERRIDE_ARG, "valid");
+  }
+  auto channel = grpc::CreateCustomChannel(server_addr, channel_creds, args);
+  std::unique_ptr<EchoTestService::Stub> stub =
+      EchoTestService::NewStub(channel);
+  EchoRequest request;
+  request.set_message("This is a test.");
+  EchoResponse reply;
+  std::cout << "Sending test message" << std::endl;
+  ClientContext context;
+  Status s = stub->Echo(&context, request, &reply);
+  EXPECT_TRUE(s.ok()) << "s.error_message() = " << s.error_message();
+}
+
+class TestServerWrapper {
+ public:
+  TestServerWrapper()
+      : server_address_("localhost:" +
+                        std::to_string(grpc_pick_unused_port_or_die())) {}
+
+  void Start(std::string certificate_file = absl::StrCat(kCredentialsDir,
+                                                         "/valid.pem"),
+             std::string key_file = absl::StrCat(kCredentialsDir, "/valid.key"),
+             std::string ca_bundle_file = absl::StrCat(kCredentialsDir,
+                                                       "/ca.pem")) {
+    auto certificate_provider =
+        std::make_shared<FileWatcherCertificateProvider>(
+            key_file, certificate_file, ca_bundle_file,
+            /*refresh_interval_sec=*/10);
+    TlsServerCredentialsOptions options(certificate_provider);
+    options.watch_root_certs();
+    options.watch_identity_key_cert_pairs();
+    options.set_cert_request_type(
+        GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY);
+    // options.set_crl_directory("");
```

Comment:
Does the problem still persist if we don't set up the CRL directory? Then probably something wrong with the client-server communication set-up. Right now we don't have the integration test for the TlsCredentials yet. Can we build this integration test on a new PR without any CRL-related changes? That can help us at least rule out if the CRL-changes are the root cause of the problem...

Reply:
One thing we can refer to is https://github.com/grpc/grpc/pull/26812/files. That PR is ongoing, but it also builds TlsCredentials in `test/cpp/end2end`, and the author said the test passed. So maybe we can take a look at that implementation as well....